crypto/modes/gcm128.c

```diff
@@ -969,8 +969,14 @@ int CRYPTO_gcm128_aad(GCM128_CONTEXT *ctx, const unsigned char *aad,
     }
 #endif
     if (len) {
+		i = 0;
         n = (unsigned int)len;
-        for (i = 0; i < len; ++i)
+        if(len >= 8)
+        {
+            ctx->Xi.u[0] ^= *(u64*)(aad);
```

Comment:
This makes an assumption that platform can tolerate unaligned references to memory. Even though most popular one does, it's not safe assumption to make. Because OpenSSL is multi-platform.

Reply:
> it would be nice to have at least one reported test for unaligned inputs.

One can pass -misalign N (up to 63) to perform benchmark with misaligned buffers. One can also pass -bytes and -seconds. Just in case for reference, remark about making assumption that platform can tolerate unaligned references to memory has nothing to do with performance, it's about portability.

> TLS 1.2 uses 13 bytes AAD.

I don't follow this remark. It's still about short TLS fragments...

> I believe that the speed reports here sensitive to the environment

And this is exactly what undermines such results. I mean it's easy to show something in benchmark, but it effectively executes in controlled environment and one can wonder how well does it reflect real life. As simplest example. The magnitude of gain here is comparable to mispredicted branch penalties. In benchmark situation the code path is practically guaranteed to execute with warmed up branch prediction. But since branch predictor is limited resource in more practical scenario chances are that it won't be warmed up. What effect would that have? So as a general rule wouldn't it make more sense to code less decisions? BTW, as for unexpected side effect. Practical value apart, if I execute speed ghash with and without suggested modification I observe up **several times drop** **with** suggested modification on my computer. Where "up to" is not only about input length, but rather ... compiler. I mean depending on compiler I use I can observe different regression coefficients. Well, to be more accurate, it's actually rather that code *with* suggested modification exhibits approximately same performance with multiple compilers, while code *without* is significantly faster and how much faster varies with compiler. Largest I measured for shortest inputs was ~2.8x difference with gcc on 12 bytes, once again, modified code is *significantly* slower on some lengths...

Reply:
VTune suggests that modified code suffers penalties from loads blocked by store forwarding [which is common when "prior store is writing a smaller region than the load is reading"]. Once again, ghash benchmark is somewhat synthetic in sense that it hardly reflects any real-life application, but the effect modification has effectively makes the point. The benchmark results should be considered with caution, especially for short inputs.